### DISCUSSION TITLE
Creating semi-complex 2D plots

### DISCUSSION THREAD

**Discussion by jvoigts** [bonsai-rx/bonsai#1066] [Q&A]
I'm working on an application where it would be extremely useful to be able to generate a real-time 2D plot of a bunch of vectors and text.
As far as I see, currently the only option is to drag points onto a canvas, which won't work due to the need to have lines, with multiple colors, arrows, dynamically updated text, and the sheer number of elements I need to plot. I attached a napkin drawing of the kind of plot I'm looking to make.

<img width="363" alt="image" src="https://user-images.githubusercontent.com/1414052/192880715-790f898a-e130-46c4-ab93-b93bbdb9903c.png">

The ideal situation as far as my particular application is concerned would be a python node that gives me access to some existing  plotting library - does this already exist or could be made with reasonable effort? I imagine there's a broader need to make real time plots of this nature to show task states, animal locations, etc. at a glance.

**Comment by glopesdev** [comment]
This is an area of fast development. In the current version you can already achieve something close to this out-of-the-box using `LineGraph`. You can draw multiple lines in 2D, control the axis range, set a rolling capacity, and even label each line individually in a legend.

It would look something like this:

<img width="375" alt="image" src="https://user-images.githubusercontent.com/5315880/193001599-c2d536eb-1415-45fe-9252-ab97c5c20b0e.png">

Attached as a zip file:
[SimpleLineGraph.zip](https://github.com/bonsai-rx/bonsai/files/9673500/SimpleLineGraph.zip)

To plot multiple line series `LineGraph` currently requires you to send pairs or tuples containing a 2D value for each series, each containing a member named X,Y. The label of each series is the name of the point-pair in the tuple, which you can set by using `ExpressionTransform`, for example like so:

Packing values into an X,Y structure:
```c#
new(
Math.Cos(it * 0.1) * 800 + 800 as X,
Math.Sin(it * 0.1) * 500 + 500 as Y)
```

Packing multiple X,Y values into a multi-series structure:
```c#
new(
Item1 as Mouse,
Item2 as Boundary)
```

To set axis boundaries in the current version you have to open the line graph visualizer, right-click on the plot and disable the auto settings and optional set the capacity. Starting from 2.7 it will be possible to set these parameters directly in the `LineGraph` node.

We are actively brainstorming future options for these nodes to make them even more flexible from within the language itself, but you can always make your own using the ZedGraph library. Here are the [docs](https://horizongir.github.io/ZedGraph/). And for examples you can look at the [ZedGraph GitHub](https://github.com/ZedGraph/ZedGraph/wiki) or maybe just look at the implementation of the [`LineGraph`](https://github.com/bonsai-rx/bonsai/blob/main/Bonsai.Design.Visualizers/LineGraph.cs) itself and surrounding classes in that folder. Feel free to drop questions in the thread, we've done pretty complex visualizers already for specific projects.

**Comment by horsto** [comment]
I want to combine multiple plots into one - mainly to save window space and to be able to control line thickness etc. with `LineGraph` controls. I found this discussion but I am already stuck because the x variable are timestamps and I want to maintain those in `LineGraph` - how do you do this?

I had a quick look at ZedGraph, but it seems even their tutorials are down / cannot be found (?). Is there any documentation written for implementation in Bonsai / a simple example?
